File: 백준/Gold/2573. 빙산/빙산.cc
#include <iostream>
#include <vector>
#include <deque>
#include <tuple>

int x_size, y_size;

std::vector<std::vector<int>> graph;
std::vector<std::vector<int>> graph2;
std::vector<std::tuple<int, int>> dx_dy_list = {std::make_tuple(-1, 0), std::make_tuple(1, 0), std::make_tuple(0, -1), std::make_tuple(0, 1)};

int dfs(int pos_x, int pos_y)
{
    int dx, dy;
    if (pos_x < 0 || pos_x >= x_size || pos_y < 0 || pos_y >= y_size)
    {
        return 0;
    }
    if (graph[pos_y][pos_x] != 0)
    {
        graph[pos_y][pos_x] = 0;
        for (std::vector<std::tuple<int, int>>::iterator it = dx_dy_list.begin(); dx_dy_list.end() != it; ++it)
        {
            auto [dx, dy] = *it;
            int new_x = pos_x + dx;
            int new_y = pos_y + dy;
            dfs(new_x, new_y);
        }
        return 1;
    }
    return 0;
}
void ice_melting(std::deque<std::tuple<int, int>> queue)
{
    int x, y, dx, dy;
    while (queue.size() != 0)
    {
        auto [x, y] = queue[0];
        queue.pop_front();
        int melting_direction_counter = 0;
        for (std::vector<std::tuple<int, int>>::iterator it = dx_dy_list.begin(); dx_dy_list.end() != it; ++it)
        {
            auto [dx, dy] = *it;
            int new_x = x + dx;
            int new_y = y + dy;
            if (new_x < 0 || new_x >= x_size || new_y < 0 || new_y >= y_size)
            {
                continue;
            }
            if (graph2[new_y][new_x] == 0)
            {
                melting_direction_counter += 1;
            }
        }
        graph[y][x] = std::max(graph2[y][x] - melting_direction_counter, 0);
    }
}

int main()
{
    std::cin >> y_size >> x_size;
    int T = 0;
    int item;
    int icebergs;
    int x, y;
    std::vector<std::tuple<int, int>> dfs_stack;
    std::deque<std::tuple<int, int>> bfs_queue;
    std::vector<int> graph_sub;
    std::vector<int> graph_sub2;
    for (int y = 0; y < y_size; y++)
    {
        for (int x = 0; x < x_size; x++)
        {
            std::cin >> item;
            graph_sub.push_back(item);
            graph_sub2.push_back(0);
        }
        graph.push_back(graph_sub);
        graph2.push_back(graph_sub2);
        graph_sub.clear();
        graph_sub2.clear();
    }
    while (1)
    {
        for (int y = 0; y < y_size; y++)
        {
            for (int x = 0; x < x_size; x++)
            {
                graph2[y][x] = graph[y][x];
            }
        }
        icebergs = 0;
        for (int y = 0; y < y_size; y++)
        {
            for (int x = 0; x < x_size; x++)
            {
                if (graph[y][x] > 0)
                {
                    dfs_stack.push_back(std::make_tuple(x, y));
                    bfs_queue.push_back(std::make_tuple(x, y));
                }
            }
        }
        for (std::vector<std::tuple<int, int>>::iterator it = dfs_stack.begin(); dfs_stack.end() != it; ++it)
        {
            auto [x, y] = *it;
            icebergs += dfs(x, y);
        }
        dfs_stack.clear();
        if (icebergs > 1)
        {
            break;
        }
        else if (bfs_queue.size() == 0)
        {
            T = 0;
            break;
        }
        else
        {
            ice_melting(bfs_queue);
            for (int y = 0; y < y_size; y++)
            {
                for (int x = 0; x < x_size; x++)
                {
                    graph2[y][x] = graph[y][x];
                }
            }
            bfs_queue.clear();
            T += 1;
        }
    }
    std::cout << T << std::endl;
}

File: 백준/Gold/2573. 빙산/빙산.py
from collections import deque
from sys import stdin, setrecursionlimit

setrecursionlimit(90000)
y_size, x_size = list(map(int, input().split(" ")))
graph: list[list[int]] = [
    list(map(int, stdin.readline().rstrip().split(" "))) for y in range(y_size)
]
graph2: list[list[int]] = [[0 for x in range(x_size)] for y in range(y_size)]
T: int = 0
dx_dy_list: list[tuple] = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def is_inbound(pos_x, x_size, pos_y, y_size):
    return 0 <= pos_x < x_size and 0 <= pos_y < y_size


def dfs(start_x, start_y):
    stack = [(start_x, start_y)]
    count = 0

    while stack:
        pos_x, pos_y = stack.pop()
        if is_inbound(pos_x, x_size, pos_y, y_size) and graph[pos_y][pos_x] != 0:
            graph[pos_y][pos_x] = 0
            count += 1
            stack.append((pos_x - 1, pos_y))
            stack.append((pos_x + 1, pos_y))
            stack.append((pos_x, pos_y - 1))
            stack.append((pos_x, pos_y + 1))

    return 1 if count > 0 else 0


# 얼음 녹이기
def ice_melting():
    while bfs_queue:
        x, y = bfs_queue.popleft()
        melting_direction_counter: int = 0
        for dx, dy in dx_dy_list:
            nx: int = x + dx
            ny: int = y + dy
            if 0 <= nx < x_size and 0 <= ny < y_size and graph2[ny][nx] == 0:
                melting_direction_counter += 1
            graph[y][x] = max(graph2[y][x] - melting_direction_counter, 0)


while True:
    graph2 = [[graph[y][x] for x in range(x_size)] for y in range(y_size)]
    icebergs: int = 0
    dfs_stack: list = []
    bfs_queue: deque = deque()
    for x in range(x_size):
        for y in range(y_size):
            if graph[y][x] > 0:
                dfs_stack.append((x, y))
                bfs_queue.append((x, y))
    # 얼음 덩어리 갯수 구하기
    for x, y in dfs_stack:
        icebergs += dfs(x, y)
    dfs_stack.clear()
    if icebergs > 1:
        break
    elif not bfs_queue:
        T = 0
        break
    else:
        ice_melting()
        graph2 = [[graph[y][x] for x in range(x_size)] for y in range(y_size)]
        T += 1
        bfs_queue.clear()
print(T)